### ISSUE TITLE
Transformer in Word Language Model predicts the same token at different steps

### ISSUE THREAD

**Issue by lileipisces** [pytorch/examples#846]
After I replaced [Line 64 in generate.py](https://github.com/pytorch/examples/blob/master/word_language_model/generate.py#L64)
`word_idx = torch.multinomial(word_weights, 1)[0]`
with
`word_idx = torch.argmax(word_weights).item()`

The model outputs

> , , , , , , , , , , , , , , , , , , , ,
>, , , , , , , , , , , , , , , , , , , ,
>, , , , , , , , , , , , , , , , , , , ,

Obviouly, it always picks the most frequently appearing token in the data as a prediction, but why? Is it because the default model size too small? Or the data too small?

**Comment by Rupeshkoushik** [comment]
Hi,i wanted to work with this issue

**Comment by lileipisces** [comment]
> Hi,i wanted to work with this issue

I have figured it out. The problem lies in the gradient clipping in [Line 181 of main.py](https://github.com/pytorch/examples/blob/master/word_language_model/main.py#L181), which is intended for RNN. After I removed this line of code, and set learning rate to 0.1, the genrated text with argmax is very fluent.

`# clip_grad_norm helps prevent the exploding gradient problem in RNNs / LSTMs.
torch.nn.utils.clip_grad_norm_(model.parameters(), args.clip)`

This issue really caused a lot of time!

**Comment by fuhaifei** [comment]
Have you found out the real cause of this mistake rather than learn rate settings? i have met the same issue.

**Comment by lileipisces** [comment]
> Have you found out the real cause of this mistake rather than learn rate settings? i have met the same issue.

I have fixed this issue, and published a paper at ACL'21. You may refer to my implementation [PETER](https://github.com/lileipisces/PETER) for the parameters settings of learning rate (1.0) and gradient clipping (1.0).

My hypothesis is that when the gradients are restricted to a very small range (0.25 in the original implementation), the search space of model parameters is small, and the model is unable to learn any pattern from the data.

**Comment by fuhaifei** [comment]
Thank you for your patient explanation. I have tried the paremeters settings  you recommended,  my model finally start to output what i want. Thank you again, It really helped me a lot.
